doc/sphinx/proofs/writing-proofs/proof-mode.rst

```diff
@@ -123,6 +123,11 @@ shows each existential with the hypotheses and conclusion for the associated goa
 
    Show Existentials.
 
+Users can control which goals are displayed in the context by :term:`focusing <focus>` and
+_shelving_ goals.  Focusing lets the user pick a single goal to work on.  Focusing operations
+can be nested.  See :ref:`focused_goals`.  Goals can be :term:`shelved` so they're not shown
+in the context until they are later :term:`unshelved`.  See :ref:`shelved_goals`.
+
```

Comment:
Q: All goals are existential variables.  Is it also true that all existential variables are goals?  I'd like to point this out or contrast them.  And IIUC different from `(exists x ...` variables.

Reply:
> Is it also true that all existential variables are goals?

Yes, it is true. Most of the time existential variables correspond to shelved goals. On the other hand, we often use the term goals to designate the goals that we work on directly, whereas shelved goals are most often resolved by a side effect.

> And IIUC different from `(exists x ...` variables.

Right. Note that a connection exists though. If you have a goal of the form `exists x, A`, then the tactic `eexists` generates a new existential variable corresponding to `x` (say `?x`) and leaves the subgoal `A`.

Reply:
> shelved goals are most often resolved by a side effect

I'm not familiar with that, that's worth describing (e.g. a paragraph + example that applies to all the e* versions of tactics).  Maybe you can provide a realistic example?  The following is probably unrealistic if not entirely on the wrong track.  I don't know if/how the proof can be completed:

```coq
Goal exists n : nat, n = n.
eexists ?[x].
  (*
  1 subgoal
  ______________________________________(1/1)
  ?x = ?x
  *)
Show Existentials.
  (*
  Existential 1 = ?x : [ |- nat] (shelved)
  Existential 2 = ?Goal : [ |- ?x = ?x]
  *)
reflexivity.
  (*
  All the remaining goals are on the shelf:

  nat
  *)
```

Reply:
This one doesn't solve the `n : nat` goal since you still need to provide a witness. What would be "solving existential by side effect" is if you had this:
```coq
Goal forall (a : nat), a = a -> exists n : nat, n = n.
Proof.
  intros a e.
  eexists ?[x].
  exact e.
Qed.
```

Now you can see that the existential variable (subgoal) n : nat can be solved by inferring from the data in `e : a = a`.

Reply:
>shelved goals are most often resolved by a side effect

Another way of explaining this is: a rule of thumb is that tactics create shelve goals by default when they correspond to existential variables that appear somewhere in one of the (non-shelved) subgoals. Similar to what the `shelve_unifiable` tactic would do. This is so because usually proving these goals will instantiate these existential variables in passing (what I call a side effect).

Reply:
This seems like a bug ("clear e" shouldn't remove e because the goal relies on it).  Right?

```coq
Goal forall (a : nat), a = a -> exists n : nat, n = n.
Proof.
intros a e.
eexists ?[x].
  (* 
    1 subgoal
    a : nat
    e : a = a
    ______________________________________(1/1)
    ?x@{a:=a; e:=e} = ?x@{a:=a; e:=e}
  *)
clear e.
  (*
    1 subgoal
    a : nat
    ______________________________________(1/1)
    ?x@{a:=a} = ?x@{a:=a}
  *)
```

How often do users need to explicitly instantiate existential variables rather than rely on side effects?

Reply:
Is there a tactic that will prove `Goal nat.`?

Reply:
>I had "Display existential variable instances" enabled in CoqIDE.

That's not something you want to activate in general. This will just print the context of the existential variable, so everything that it *could* eventually depend on when instantiated. However, this does not make the goal depend on all of this context and thus the behavior of `clear` is perfectly OK here.

>How often do users need to explicitly instantiate existential variables rather than rely on side effects?

Rarely.



> Is there a tactic that will prove `Goal nat.`?
> 
> I guess that's just `apply O` or `apply (S (O))` ...

Yep. Or just `exact 0`.